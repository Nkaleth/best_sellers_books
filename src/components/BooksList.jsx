import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { getBestSellerBooks } from '../redux/details/detailsSlice';
import Navbar from './Navbar';
import Book from './Book';
import '../styles/booklist.css';

const BooksList = ({ id }) => {
  const {
    booksDetails, listName, isLoading, publishedDate,
  } = useSelector((store) => store.details);
  const itemUrl = `https://api.nytimes.com/svc/books/v3/lists/${id}.json?api-key=ojoeVThZ3GLU2SAZCITvtT1smz6qZuGq`;
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getBestSellerBooks(itemUrl));
  }, [itemUrl, dispatch]);
  if (isLoading) {
    return (
      <div className="loading">
        <h1>Loading...</h1>
      </div>
    );
  }

  return (
    <>
      <Navbar categorie={listName} numCateg={booksDetails.length} />
      <p className="listDate">{`Ranking Date: ${publishedDate}`}</p>
      <section className="booksDetailsBox">
        { booksDetails.map((book) => (
          <Book
            key={book.primary_isbn10}
            rank={book.rank}
            rankLastWeek={book.rank_last_week}
            weeksOnList={book.weeks_on_list}
            title={book.title}
            author={book.author}
            bookImg={book.book_image}
          />
        ))}
      </section>
    </>
  );
};

BooksList.propTypes = { id: PropTypes.string.isRequired };

export default BooksList;
