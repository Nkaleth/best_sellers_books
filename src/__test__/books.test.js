import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import BooksList from '../components/BooksList';

describe('Tests for details page: <BooksList /> component', () => {
  const initialState = {
    booksDetails: [
      {
        rank: 1,
        rank_last_week: 15,
        weeks_on_list: 9,
        title: 'THE MAID',
        author: 'Nita Prose',
        book_image: 'https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg',
      },
      {
        rank: 2,
        rank_last_week: 15,
        weeks_on_list: 9,
        title: 'Book 2',
        author: 'Author2',
        book_image: 'https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg',
      },
      {
        rank: 3,
        rank_last_week: 15,
        weeks_on_list: 9,
        title: 'Book 3',
        author: 'Author3',
        book_image: 'https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg',
      },
      {
        rank: 4,
        rank_last_week: 15,
        weeks_on_list: 9,
        title: 'Book 4',
        author: 'Author4',
        book_image: 'https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg',
      },
    ],
    listName: 'No list loaded',
    publishedDate: 'no Date',
    isLoading: false,
  };

  const detailsSlice = createSlice({
    name: 'details',
    initialState,
  });

  const store = configureStore({
    reducer: {
      details: detailsSlice.reducer,
    },
  });

  const wrapper = (component) => (
    <Provider store={store}>
      <BrowserRouter>
        {component}
      </BrowserRouter>
    </Provider>
  );

  it("Should render '4 categories' text", () => {
    render(wrapper(<BooksList />));
    expect(screen.getByText(/4 categories/i)).toBeInTheDocument();
  });

  it('Should render 4 books', () => {
    render(wrapper(<BooksList />));
    expect(screen.getAllByTestId('bookBox')).toHaveLength(4);
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(<BooksList />)).toJSON()).toMatchSnapshot();
  });
});
