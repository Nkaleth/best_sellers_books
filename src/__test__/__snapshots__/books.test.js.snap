// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing details page: <BooksList /> component Should match the snapshot 1`] = `
Array [
  <nav
    className="navbarTop"
  >
    <div
      className="contentnavbarTop"
    >
      <a
        aria-current="page"
        className="iconBox active"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="goBack"
          src="backward-step-svgrepo-com.svg"
        />
        <p
          className="yearNavbar"
        >
          2023
           
        </p>
      </a>
      <p
        className="titleNavbar"
      >
        New York Times Best Sellers Books
      </p>
      <div
        className="iconBox"
      >
        <img
          alt="mic"
          src="mic-svgrepo-com.svg"
        />
      </div>
      <div
        className="iconBox"
      >
        <img
          alt="config"
          src="gear-svgrepo-com.svg"
        />
      </div>
    </div>
  </nav>,
  <section
    className="navbarBottom"
  >
    <div
      className="catImgContainer"
    >
      <img
        alt="catImg"
        src="library-svgrepo-com.svg"
      />
    </div>
    <div
      className="titleCountBox"
    >
      <h1
        className="titleCategories"
      >
        No list loaded
      </h1>
      <p
        className="counterNavbar"
      >
        4
         
        Items
      </p>
    </div>
  </section>,
  <p
    className="listDate"
  >
    Ranking Date: no Date
  </p>,
  <section
    className="booksDetailsBox"
  >
    <article
      className="bookBox"
      data-testid="bookBox"
    >
      <section
        className="bookTitle"
      >
        <div
          className="rankData"
        >
          <h2
            className="rankNumber"
          >
            1
          </h2>
          <div
            className="arrowRankBox"
          >
            <img
              alt="UpArrow"
              className="arrowRank"
              src="up-arrow-svgrepo-com.svg"
            />
          </div>
          <div
            className="bookCoverBox"
          >
            <img
              alt="THE MAID"
              src="https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg"
            />
          </div>
        </div>
        <div
          className="titleAuthorBook"
        >
          <h2>
            THE MAID
          </h2>
          <span>
            by Nita Prose
          </span>
          <span
            className="weeksList"
          >
            9 WEEKS ON THE LIST
          </span>
        </div>
      </section>
    </article>
    <article
      className="bookBox"
      data-testid="bookBox"
    >
      <section
        className="bookTitle"
      >
        <div
          className="rankData"
        >
          <h2
            className="rankNumber"
          >
            2
          </h2>
          <div
            className="arrowRankBox"
          >
            <img
              alt="UpArrow"
              className="arrowRank"
              src="up-arrow-svgrepo-com.svg"
            />
          </div>
          <div
            className="bookCoverBox"
          >
            <img
              alt="Book 2"
              src="https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg"
            />
          </div>
        </div>
        <div
          className="titleAuthorBook"
        >
          <h2>
            Book 2
          </h2>
          <span>
            by Author2
          </span>
          <span
            className="weeksList"
          >
            9 WEEKS ON THE LIST
          </span>
        </div>
      </section>
    </article>
    <article
      className="bookBox"
      data-testid="bookBox"
    >
      <section
        className="bookTitle"
      >
        <div
          className="rankData"
        >
          <h2
            className="rankNumber"
          >
            3
          </h2>
          <div
            className="arrowRankBox"
          >
            <img
              alt="UpArrow"
              className="arrowRank"
              src="up-arrow-svgrepo-com.svg"
            />
          </div>
          <div
            className="bookCoverBox"
          >
            <img
              alt="Book 3"
              src="https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg"
            />
          </div>
        </div>
        <div
          className="titleAuthorBook"
        >
          <h2>
            Book 3
          </h2>
          <span>
            by Author3
          </span>
          <span
            className="weeksList"
          >
            9 WEEKS ON THE LIST
          </span>
        </div>
      </section>
    </article>
    <article
      className="bookBox"
      data-testid="bookBox"
    >
      <section
        className="bookTitle"
      >
        <div
          className="rankData"
        >
          <h2
            className="rankNumber"
          >
            4
          </h2>
          <div
            className="arrowRankBox"
          >
            <img
              alt="UpArrow"
              className="arrowRank"
              src="up-arrow-svgrepo-com.svg"
            />
          </div>
          <div
            className="bookCoverBox"
          >
            <img
              alt="Book 4"
              src="https://storage.googleapis.com/du-prd/books/images/9780593356159.jpg"
            />
          </div>
        </div>
        <div
          className="titleAuthorBook"
        >
          <h2>
            Book 4
          </h2>
          <span>
            by Author4
          </span>
          <span
            className="weeksList"
          >
            9 WEEKS ON THE LIST
          </span>
        </div>
      </section>
    </article>
  </section>,
]
`;
